aws-datastore/src/main/java/com/amplifyframework/datastore/network/SubscriptionProcessor.java

```diff
@@ -0,0 +1,126 @@
+import com.amplifyframework.datastore.DataStoreChannelEventName;
+import com.amplifyframework.datastore.DataStoreException;
+import com.amplifyframework.datastore.storage.LocalStorageAdapter;
+import com.amplifyframework.datastore.storage.StorageItemChange;
+import com.amplifyframework.hub.HubChannel;
+import com.amplifyframework.hub.HubEvent;
+import com.amplifyframework.logging.Logger;
+
+import io.reactivex.Single;
+import io.reactivex.disposables.CompositeDisposable;
+import io.reactivex.subjects.ReplaySubject;
+
+/**
+ * Starts subscriptions to AppSync. Applies data to local storage when it arrives.
+ *
+ * TODO: this component should save data via the merger, not directly through the
+ * {@link LocalStorageAdapter}.
+ */
+final class SubscriptionProcessor {
+    private static final Logger LOG = Amplify.Logging.forNamespace("amplify:aws-datastore");
+    private final RemoteModelMutations remoteModelMutations;
+    private final LocalStorageAdapter localStorageAdapter;
+    private final CompositeDisposable disposable;
+    private final ReplaySubject<Mutation<? extends Model>> buffer;
+
+    SubscriptionProcessor(
+            @NonNull LocalStorageAdapter localStorageAdapter,
+            @NonNull AppSyncEndpoint appSyncEndpoint,
+            @NonNull ModelProvider modelProvider) {
+        this.localStorageAdapter = localStorageAdapter;
+        this.remoteModelMutations = new RemoteModelMutations(appSyncEndpoint, modelProvider);
+        this.disposable = new CompositeDisposable();
+        this.buffer = ReplaySubject.create();
+    }
+
+    /**
+     * Start subscribing to model mutations.
+     */
+    void startSubscriptions() {
+        disposable.add(remoteModelMutations.observe()
+            .subscribe(
+                mutationOnSubscription -> {
+                    buffer.onNext(mutationOnSubscription);
+                    LOG.info("Successfully enqueued mutation from subscription: " + mutationOnSubscription);
+                },
+                error -> LOG.warn("Error enqueuing mutation from subscription.", error),
+                () -> LOG.warn("Subscription to remote model mutations is completed.")
+            )
+        );
+    }
+
+    /**
+     * Start draining mutations out of the mutation buffer.
+     * This should be called after {@link #startSubscriptions()}.
+     */
+    void startDrainingMutationBuffer() {
+        disposable.add(buffer
+            .flatMapSingle(this::applyMutationToLocalStorage)
+            .subscribe(
+                savedMutation -> LOG.info("Saved a mutation from a subscription: " + savedMutation),
+                failure -> LOG.warn("Reading subscriptions buffer has failed.", failure),
+                () -> LOG.warn("Reading from subscriptions buffer is completed.")
+            )
+        );
+    }
+
+    /**
+     * Stop any active subscriptions, and stop draining the mutation buffer.
+     */
+    void stopAllSubscriptionActivity() {
+        disposable.clear();
```

Comment:
should we also clear `buffer` here?

Reply:
The thing doing the buffering is the `ReplaySubject`; this `disposable` represents a subscription to that subject. In Rx1, they called the `Disposable` as `Subscription`, which I liked better. But anyway, TLDR, yea I'd be open to a better name for this ... maybe `subscriptionToBuffer`, instead of the (what does it mean??) `disposable`?